src/WireBST.java

```diff
@@ -109,29 +109,242 @@ private void insertToRight(WireBSTNode currentNode , WireBSTNode newNode ){
-        if(newNode.getStudentNum() > this.getMedianWireBST().getStudentNum() ){
-            if (this.getNodeCounter() % 2 == 0){
-                this.setMedian ( getSuccessor( this.getMedianWireBST() ) ) ;
-            }
-        else
-            if (this.getNodeCounter() % 2 == 1){
-                this.setMedian ( getPredecessor( this.getMedianWireBST() ) ) ;
+        if(newNode.getStudentNum() > this.getMedianWireBST().getStudentNum() ) {
+            if (this.getNodeCounter() % 2 == 0) {
+                this.setMedian(getSuccessor(this.getMedianWireBST()));
             }
+        }
+        else if (this.getNodeCounter() % 2 == 1) {
+            this.setMedian(getPredecessor(this.getMedianWireBST()));
+        }
         this._nodeCounter ++;
+    }
+
+    /**
+     * remove the student that is number is the given parameter
+     * @param studentNum, the node who include that number
+     * @return true if the node was removed, else false
+     */
+    public boolean removeWireBSTNode(int studentNum){
+
+        WireBSTNode temp = searchWireBST( studentNum );
+
+        if (temp == null){
+
+            return false;
+        }
+
+        else {
+
+           removeWireBSTNode(temp);
+           return true;
+
+        }
+    }
+
+    /**
+     * overloading the removeWireBSTNode function, remove node who exist in the tree
+     * @param node is the node to remove.
+     */
+    public void removeWireBSTNode(WireBSTNode node){
+
+        if ( ! node.isRealLeft() && ! node.isRealRight() ){
+
+            removeDeepLeaf(node);
+        }
+
+        if ((node.isRealRight()) && (!node.isRealLeft())){    //** in case of removing node who has only right son
+
+            removeNodeWithOnlyRightSon(node);
+
+        }
+
+        if ((!node.isRealRight()) && (node.isRealLeft())){    //** in case of removing node who has only left son
+
+            removeNodeWithOnlyLeftSon(node);
+
+        }
+
+        if ((node.isRealRight()) && (node.isRealLeft())) {    //** in case of removing node who has two sons
+
+            removeNodeWithTwoSons(node);
+
+        }
+    }
+
+    /**
+     * removing a deep leaf. used by the removeWiredBSTNode function
+     * @param node the node we wish to remove
+     */
+    private void removeDeepLeaf(WireBSTNode node){
+
+        if (node.getParent() == null){                              //in case node is root and has no sons
+
+            this.setHead(null);
+            this.setMedian(null);
+
+        }
+
+        else if (node.equal(node.getParent().getRight())) {
+
+            updateMedianRemove(node);
+            node.getParent().setRight(getSuccessor(node));
+        }
+
+        else {
+
+            updateMedianRemove(node);
+            node.getParent().setLeft(getPredecessor(node));
+
+        }
+    }
+
+    /**
+     * update the median after remove of a node
+     * @param node, the node that removed
+     */
+    private void updateMedianRemove(WireBSTNode node){
+
```

Comment:
if( this.getCounter() == 1 )
this.setMedian(null);

Reply:
done